`skywire-cli config priv` subcommand

0pcom opened the pull request:
Did you run `make format && make check`?
yes

Fixes 
#1361 
#1362 

 Adds:	
- `skywire-cli config priv` subcommands to `set` and `get` rewards address and privacy with privacy.json
- `skywire-cli visor priv` subcommands for setting privacy over rpc
- SetPrivacy and GetPrivacy rpc endpoints
- system hardware survey generation (system.json)

How to test this PR:
```
skywire-cli config priv --help
skywire-cli config priv set -- help
skywire-cli config priv get -- help
skywire-cli config priv set
skywire-cli config priv get
skywire-cli config priv set <skycoin-address>
skywire-cli config priv set -a <skycoin-address> -i -o $(pwd)
skywire-cli visor priv set
skywire-cli visor priv get
```

Commit:
add priv subcommand to skywire-cli config for generating privacy.json file

Commit:
add cmd/skywire-cli/commands/config/private_linux.go

Commit:
improve error checking ; streamline check for root permissions for config gen & priv

Commit:
make format check

Commit:
add skywire-cli visor logserver subcommand for serving logs via dmsghttp

Commit:
fix appveyor windows errors

Commit:
add logserver to visor ; not working right yet. Added -x flag to visor for disabling the hypervisor as a testing consideration.

Commit:
partial update to api to support generating the privacy.json

Commit:
Merge branch 'develop' into feat/privacyjson

Commit:
add getprivacy & setprivacy api endpoints. add skywire-cli visor priv subcommand. Add get and set subcommands to priv subcommands in skywire-cli visor and skywire-cli config. move Privacy struct to skyenv.

Commit:
add skywire-cli visor priv subcommand

Commit:
add system hardware survey

Commit:
remove nonworking dmsghttp logserver module from visorinit

Commit:
add hypervisor api endpoints for checking and setting privacy

jdknives commented on cmd/skywire-cli/commands/visor/logserver.go:
```diff
@@ -0,0 +1,98 @@
+	Use:   "logserver",
+	Short: "log server",
+	Long: `dmsghttp log server
+
```
Formatting seems off in source. Maybe this looks different in print on the CLI?

0pcom replied:
Initially, I started with copying the example dmsghttp server from the dmsg repo into a cli command, with the intent of running it as its own service independent of the visor.

I realized this would not work because we would need a similar non-systemd framework for this service on windows and macOS as we want to collect logs for transports from vpn clients as well as servers for transport log reconciliation.

So I wrote the logserver into the visor, but then removed it as it was not working correctly and erson had it working in his pr.

logserver.go is the initial effort in this direction which was ultimately abandoned and will be removed with the next commit

jdknives commented on cmd/skywire-cli/commands/visor/logserver.go:
```diff
@@ -0,0 +1,98 @@
+	dir            = skyenv.PackageConfig().LocalPath // local dir to serve via http
+	dmsgDisc       = "http://dmsgd.skywire.skycoin.com"
+	dmsgPort       = uint(81)
+	pubkey, seckey = cipher.GenerateKeyPair() //nolint
```
What is this supposed to do? If we generate a new keypair here instead of using the visors keypair, we will never be able to retrieve these logs in practice. We want to fetch logs from a given PubKey, how would we know what PubKey the corresponding logserver has here?

jdknives replied:
After reading the code on line https://github.com/skycoin/skywire/pull/1369/files#diff-c954be1d1e9b19b826bd63f7a85502538af6fa502ff0eed1915385db6ec3c5c2R59 I dont get why we initialize these variables here at all.

jdknives commented on cmd/skywire-cli/commands/visor/logserver.go:
```diff
@@ -0,0 +1,98 @@
+)
+
+var (
+	dir            = skyenv.PackageConfig().LocalPath // local dir to serve via http
```
Maybe have a separate directory inside the local dir instead of just serving the whole local dir with additional files we place there.

jdknives commented on cmd/skywire-cli/commands/visor/logserver.go:
```diff
@@ -0,0 +1,98 @@
+
+var (
+	dir            = skyenv.PackageConfig().LocalPath // local dir to serve via http
+	dmsgDisc       = "http://dmsgd.skywire.skycoin.com"
```
I guess this may be due to you wanting to have something you can test with quick, but why are we not using the discovery set in config?

jdknives commented on cmd/skywire-cli/commands/visor/logserver.go:
```diff
@@ -0,0 +1,98 @@
+	Run: func(cmd *cobra.Command, args []string) {
+		mLog := logging.NewMasterLogger()
+		mLog.SetLevel(logrus.InfoLevel)
+		log := logging.MustGetLogger("dmsghttp-logserver")
```
Not sure if its important to know for users/log output that this module is running over dmsghttp. Its just a logserver.

jdknives commented on cmd/skywire-cli/commands/visor/logserver.go:
```diff
@@ -0,0 +1,98 @@
+		log := logging.MustGetLogger("dmsghttp-logserver")
+		ctx, cancel := cmdutil.SignalContext(context.Background(), log)
+		defer cancel()
+		if !skyenv.IsRoot() {
```
Why does it need to run as root/

jdknives commented on cmd/skywire-cli/commands/visor/logserver.go:
```diff
@@ -0,0 +1,98 @@
+		ctx, cancel := cmdutil.SignalContext(context.Background(), log)
+		defer cancel()
+		if !skyenv.IsRoot() {
+			log.Fatal("Log server is designed to run as root.")
```
Maybe saying something like `Log server requires root permissions` would be more explicit if we indeed need root permissions.

jdknives commented on cmd/skywire-cli/commands/visor/logserver.go:
```diff
@@ -0,0 +1,98 @@
+		if !skyenv.IsRoot() {
+			log.Fatal("Log server is designed to run as root.")
+		}
+		log.WithField("config filepath", skyenv.SkywirePath+"/"+skyenv.Configjson).Info()
```
What is the purpose of this log at `INFO` level? Maybe helpful for you as you test this server?

jdknives commented on cmd/skywire-cli/commands/visor/privacy.go:
```diff
@@ -0,0 +1,87 @@
+		if err != nil {
+			log.WithError(err).Fatal("Could not unmarshal json.")
+		}
+		fmt.Printf("%s", j)
```
This should also print a newline even if the file does not end on one, right?

0pcom replied:
I believe valid files in linux contain a trailing newline

in my recent comment I have documented the command outputs and there is a trailing newline

jdknives commented on cmd/skywire-visor/commands/root.go:
```diff
@@ -97,6 +92,8 @@ func init() {
 		rootCmd.Flags().BoolVarP(&launchBrowser, "browser", "b", false, "open hypervisor ui in default web browser")
 	}
 	rootCmd.Flags().BoolVarP(&hypervisorUI, "hvui", "i", false, "run as hypervisor")
+	rootCmd.Flags().BoolVarP(&noHypervisorUI, "nohvui", "x", false, "disable hypervisor")
```
Whats this doing now? Ad-hoc disabling the hypervisorUI? I guess this is useful for you somehow...

0pcom replied:
Yes, i needed that for testing purposes.

jdknives commented on pkg/skyenv/values.go:
```diff
@@ -170,3 +173,44 @@ func HomePath() string {
+const SurveyFile string = "system.json"
+
+// HwSurvey returns system hardware survey
+func HwSurvey() (s Survey) {
```
Maybe call this System Survey? Not sure if I named it well in the Specs, but Synth typically calls it system survey anyways.

0pcom replied:
will change on next commit

jdknives commented on pkg/skyenv/values.go:
```diff
@@ -170,3 +173,44 @@ func HomePath() string {
+
+// Config returns either UserConfig or PackageConfig based on permissions
+func Config() PkgConfig {
+	if IsRoot() {
```
I dont understand why we are returning different configs based on permissions. It sounds like we assume that if the binary is run as root, we are dealing with a package install, which may be the case reasonably often, but will frustrate anyone not using the package install with root privileges also running the VPN client with root.

0pcom replied:
The permission tells where files should be written to, basically. So you don't want to have the visor use a config with paths that it either can't write to or shouldn't write too

This was basically existing logic that i moved from a few different places to here and it's not universally implemented, only where appropriate.

jdknives commented on pkg/visor/api.go:
```diff
@@ -303,6 +307,40 @@ func (v *Visor) Uptime() (float64, error) {
+ */
+
+// SetPrivacy implements API.
+func (v *Visor) SetPrivacy(p skyenv.Privacy) error {
```
Not sure if its a wise move to make the visor binary dependent on the existence of the `skywire-cli` command being existent and in $PATH. Why not write to the config yourself here?

jdknives replied:
Also, it appears to me like we are calling `SetPrivacy` function inside the `skywire-cli config priv set` call via RPC. And then inside this function we are calling the cli function again. Are we sure this works.

0pcom replied:
it is my strong conviction that the visor should be writing its own configs. The cli is tasked with config generation and updating.

the issue we have had previously is with having multiple implementations of the same thing, and a certain bifurcation or lack of feature parity existed in that.

with this, the config is only written by the cli, and everything else calls upon that existing functionality.

the visor is, in a practical sense, dependant on skywire-cli. We don't explicitly support omitting skywire-cli from the installation. I don't know how you could practically or reliably run skywire without updating the config on version changes, that sort of thing. You would need skywire-cli at some point and there is no reason to separate the binaries.

lets please assume it's in the path and not go down such rabbit holes. If the cli cannot be assumed to be present with the visor i will begin to question my sanity.

it's not like it breaks the visor if there is a command not found error there.

ersonp replied:
The way the code is written right now is overly complicated and the API part of it won't even work because it simply calls itself. I think there is a way to fix it by simply moving the code that writes to the `privacy.json` to internal or to it's own pkg and that way both the instances will be using the exact same code.

Also, there is no need to use the cli in this way since we are in the same codebase and can access the code programmatically. Because of the drawbacks of not having cli in the PATH using the current way may not cause any issues when installed by packages but will most certainly cause unnecessary headaches and issues while development and testing and also on a manual install/build.

Commit:
fix GetPrivacy and SetPrivacy in visor api / rpc / rpc_client

Commit:
add cmd/skywire-cli/commands/config/private.go

Commit:
git add -f vendor/*

0pcom commented:
```
$ skywire-cli config priv --help
rewards & privacy setting

Sets the skycoin rewards address and ip public for the visor.
The config is written to the root of the default local directory
Run this command with root permissions for visors running as root via systemd
this config is served via dmsghttp along with transport logs
and the system hardware survey for automating rewards distribution

Usage:
  skywire-cli config priv [command]

Available Commands:
  set          set reward address & node privacy
  get          read reward address & privacy setting from file

Global Flags:
      --json   print output in json

Use "skywire-cli config priv [command] --help" for more information about a command.
```
The "default" (for testing) is the genesis address
```
$ skywire-cli config priv set --help
set reward address & node privacy

Usage:
  skywire-cli config priv set <address> [flags]

Flags:
  -a, --address string   reward address (default "2jBbGxZRGoQG1mqhPBnXnLTxK6oxsTf8os6")
  -o, --out string       output config: /opt/skywire/local/privacy.json
  -i, --publicip         display node ip

Global Flags:
      --json   print output in json
```
```
$ skywire-cli config priv get --help
read reward address & privacy setting from file

Usage:
  skywire-cli config priv get [flags]

Flags:
  -o, --out string   read config from: /opt/skywire/local/privacy.json

Global Flags:
      --json   print output in json
```
```
$ sudo skywire-cli config priv set
[2022-10-01T17:56:48-05:00] INFO [skywire-cli]: Updated file '/opt/skywire/local/privacy.json' to:
{
	"display_node_ip": false,
	"reward_address": "2jBbGxZRGoQG1mqhPBnXnLTxK6oxsTf8os6"
}
```
```
$ skywire-cli config priv get
{
	"display_node_ip": false,
	"reward_address": "2jBbGxZRGoQG1mqhPBnXnLTxK6oxsTf8os6"
}
```
demonstrating using `skycoin-cli` to get an address from a wallet to set reward address
```
$ sudo skywire-cli config priv set $(skycoin-cli listAddresses $HOME/.skycoin/wallets/*.wlt | head -n3 | tail -n1 | tr -d '"')
[2022-10-01T17:56:48-05:00] INFO [skywire-cli]: Updated file '/opt/skywire/local/privacy.json' to:
{
	"display_node_ip": false,
	"reward_address": "nLpzxwcH78ytc6k3ff1SfBrVCQHKaZ59SD"
}
```
demonstrating setting a custom path and setting display_node_ip true
```
$skywire-cli config priv set -a $(skycoin-cli listAddresses $HOME/.skycoin/wallets/*.wlt | head -n3 | tail -n1 | tr -d '"') -i -o $(pwd)/privacy.json
[2022-10-01T17:56:48-05:00] INFO [skywire-cli]: Updated file '/home/user/go/src/github.com/skycoin/skywire/privacy.json' to:
{
	"display_node_ip": true,
	"reward_address": "nLpzxwcH78ytc6k3ff1SfBrVCQHKaZ59SD"
}
```
the user will either set the privacy.json file via `skywire-cli config priv set` or from the hypervisor UI. To make the configuration possible, rpc functions were added.  `skywire-cli visor priv set` & `get` are more or less a test of the rpc.

the underlying command executed is `skywire-cli config priv set` or `get`

config generation is intentionally done exclusively via skywire-cli, as it is not appropriate for the visor to generate or update configs except for very limited circumstances such as updating app autostart via the hypervisor UI
```
$ skywire-cli visor priv set --help
configure privacy settings

test of the api endpoint SetPrivacy

Usage:
  skywire-cli visor priv set [flags]

Flags:
  -a, --address string   reward address (default "2jBbGxZRGoQG1mqhPBnXnLTxK6oxsTf8os6")
  -o, --out string       output config: /home/user/.skywire/local/privacy.json
  -i, --publicip         display node ip

Global Flags:
      --json         print output in json
      --rpc string   RPC server address (default "localhost:3435")
```

`skywire-cli visor priv get` will show the contents of privacy.json if it exists in the visor's local path. By contrast `skywire-cli config priv get` reads the config from the default local path the visor uses when run as root / default for package installation.
```
$ skywire-cli visor priv get --help
configure privacy settings

test of the api endpoints GetPrivacy

Usage:
  skywire-cli visor priv get [flags]

Global Flags:
      --json         print output in json
      --rpc string   RPC server address (default "localhost:3435")
```

This seems like it should create the config with the wrong permissions for the path but in fact it does not
```
$ skywire-cli visor priv set -a $(skycoin-cli listAddresses $HOME/.skycoin/wallets/*.wlt | head -n3 | tail -n1 | tr -d '"') -i -o $(pwd)/privacy.json
[2022-10-01T17:56:48-05:00] INFO [skywire-cli visor priv set]: Privacy settings updated to:
{"display_node_ip":true,"reward_address":"nLpzxwcH78ytc6k3ff1SfBrVCQHKaZ59SD"}
```

```
$ skywire-cli visor priv get
{
	"display_node_ip": true,
	"reward_address": "nLpzxwcH78ytc6k3ff1SfBrVCQHKaZ59SD"
}
```
the config was removed here without permissions errors
```
$ cat privacy.json
{
	"display_node_ip": true,
	"reward_address": "nLpzxwcH78ytc6k3ff1SfBrVCQHKaZ59SD"
}
$ rm privacy.json

```

ersonp committed:
Merge branch 'develop' into feat/privacyjson

ersonp committed:
Rename HwSurvey to SystemSurvey
